survivalstan/formulas.py

```diff
@@ -0,0 +1,414 @@
+    ''' Class of errors pertaining to invalid Id variables '''
+
+class Surv(object):
+    ''' Class representing stateful Survival-type data
+    '''
+    def __init__(self):
+        self.subject_id = Id('subject')
+        self.timepoint_id = Id('timepoint')
+        self.group_id = Id('group')
+        self._type = None
+        self._grouped = None
+        pass
+
+    def _check_kwargs(self, **kwargs):
+        kwargs = dict(**kwargs)
+        allowed_kwargs = ['subject', 'group']
+        bad_keys = [key not in allowed_kwargs for key in kwargs.keys()]
+        if any(bad_keys):
+            raise ValueError('Invalid parameter: {}'.format(','.join(bad_keys)))
+        return kwargs
+
+    def memorize_chunk(self, time, event_status, **kwargs):
+        kwargs = self._check_kwargs(**kwargs)
+        if 'subject' in kwargs.keys():
+            self._type = 'long'
+            self.subject_id.memorize_chunk(kwargs['subject'])
+            self.timepoint_id.memorize_chunk(time)
+        else:
+            self._type = 'wide'
+        if 'group' in kwargs.keys():
+            self._grouped = True
+            self.group_id.memorize_chunk(kwargs['group'])
+        else:
+            self._grouped = False
+
+    def memorize_finish(self):
+        self.subject_id.memorize_finish()
+        self.group_id.memorize_finish()
+        self.timepoint_id.memorize_finish()
+
+    def _prep_timepoint_standata(self, timepoint_df):
+        unique_timepoints = _prep_timepoint_dataframe(
+            timepoint_df,
+            timepoint_id_col='id',
+            timepoint_end_col='value')
+        timepoint_input_data = {
+            't_dur': unique_timepoints['t_dur'],
+            't_obs': unique_timepoints['value'],
+            'T': len(unique_timepoints.index)
+        }
+        return timepoint_input_data
+
+    def _prep_long(self, timepoint_id, event_status, subject_id, group_id=None,
+                   stan_data=dict(), meta_data=dict(), **kwargs):
+        if patsy.util.have_pandas:
+            dm = {'timepoint_id': timepoint_id,
+                  'event_status': event_status.astype(int),
+                  'subject_id': subject_id
+                  }
+            if group_id is not None:
+                dm.update({'group_id': group_id})
+            dm = pd.DataFrame(dm)
+            # prep stan_data inputs
+            stan_data.update({
+                    'event': dm['event_status'].values.astype(int),
+                    't': dm['timepoint_id'].values.astype(int),
+                    's': dm['subject_id'].values.astype(int),
+                    'N': len(dm.index),
+                    })
+            if group_id is not None:
+                stan_data.update({'g': dm['group_id'].values.astype(int)})
+            meta_data.update({'df': dm})
+        else:
+            raise ValueError('non-pandas use case not supported yet. Please ',
+                             'import pandas to use `surv`')
+            #if group_id is not None:
```

Comment:
I originally didn't have the `ValueError` there, but then I noticed I couldn't do any tests without pandas since `SurvivalStan` depends on pandas. I added the `ValueError` precisely so that the non-tested code wouldn't run.

However, almost all the `patsy` extensions support a with-pandas & a without-pandas use case.

What do you suggest - should i just remove the following lines? They would be useful if I separated this `surv` functionality from `survivalstan`, but are not useful now. 

Thank you thank you for your feedback